### ISSUE TITLE
Buffering

### ISSUE THREAD

**Issue by GoogleCodeExporter** [blyzer/morelinq#87]
```
The use of buffering on Repeat() and Cartesian() is currently commented as a 
TODO. The attached patch:

1. Implements a buffering class that buffers enumerables in memory.
2. Changes Repeat() and Cartesian to use this buffering.
3. Further changes Repeat() to skip this if the count is 0 or 1.
4. Defines an IBufferedEnumerable<T> interfaces, documented as adding a 
buffering guarantee to IEnumerable<T>.
5. Defines two overloads of a method to create such a buffered instance. One of 
which insists that buffering happen even if the source implements 
ICollection<T>. (Otherwise it is assumed that the source already serves as a 
buffer adequately).

In short: Deals with the two TODO items, and brings the advantage of the code 
needed to do so, to the user.
```

Original issue reported on code.google.com by `j...@hackcraft.net` on 12 Dec 2013 at 2:46

Attachments:
- [Buffering.patch](https://storage.googleapis.com/google-code-attachments/morelinq/issue-87/comment-0/Buffering.patch)

**Comment by GoogleCodeExporter** [comment]
```
Sorry, there was a slight error in the XML documentation. Further patch:
```

Original comment by `j...@hackcraft.net` on 12 Dec 2013 at 2:52

Attachments:
- [Buffering2.patch](https://storage.googleapis.com/google-code-attachments/morelinq/issue-87/comment-1/Buffering2.patch)

**Comment by GoogleCodeExporter** [comment]
```
This issue has been migrated to:
https://github.com/MoreLINQ/morelinq/issues/87
The conversation continues there.
DO NOT post any further comments to the issue tracker on Google Code as it is 
shutting down.
You can also just subscribe to the issue on GitHub to receive notifications of 
any further development.
```

Original comment by `azizatif` on 21 Aug 2015 at 6:56
